Data/Ring/Ordered.hs

```diff
@@ -0,0 +1,377 @@
+{-# LANGUAGE CPP                        #-}
+{-# LANGUAGE DeriveDataTypeable         #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE StandaloneDeriving         #-}
+{-# LANGUAGE Trustworthy                #-}
+
+-- |
+-- Module: Data.Ring.Ordered
+-- Copyright: (C) 2021 Koz Ross
+-- License: 
+-- Maintainer: Koz Ross <koz.ross@retro-freedom.nz>
+-- Stability: stable
+-- Portability: GHC only
+--
+-- An \'ordered ring\' is a ring with a total order.
+--
+-- = Mathematical pedantry note
+--
+-- Many (if not most) of the instances of the 'OrderedRing' type class are not
+-- truly ordered rings in the mathematical sense, as the
+-- [axioms](https://en.wikipedia.org/wiki/Ordered_ring) imply that the
+-- underlying set is either a singleton or infinite. Thus, the [additional
+-- properties](https://en.wikipedia.org/wiki/Ordered_ring#Basic_properties) of
+-- ordered rings do not, in general, hold. The only exception is that for any
+-- @OrderedRing a => x :: a@, @x@ is either positive (that is, @'signum' x ==
+-- 'Positive'@), negative (that is, @'signum' x == 'Negative'@) or @zero@.
+--
+-- We indicate those instances that /are/ \'truly\' or \'mathematically\'
+-- ordered rings in their documentation.
+module Data.Ring.Ordered 
+  (
+    -- * Helper types
+    Signum(..), 
+    Modular(..),
+    -- * Ordered ring type class
+    OrderedRing(..),
+  ) where
+
+import Control.Applicative (Const (Const))
+#if MIN_VERSION_base(4,7,0)
+import Data.Data (Data)
+#endif
+import Data.Fixed (HasResolution, Fixed)
+import Data.Functor.Identity (Identity (Identity))
+import Data.Int (Int8, Int16, Int32, Int64)
+import Data.Monoid (Dual (Dual))
+import Data.Ord (Down (Down))
+import Data.Ratio (Ratio)
+import Data.Semiring (Ring (negate), Semiring(zero, one))
+import Data.Semiring.Generic ()
+import Data.Word (Word8, Word16, Word32, Word64)
+import GHC.Generics (Generic)
+import Prelude hiding (signum, abs, negate, (-))
+import qualified Prelude as Num
+#if MIN_VERSION_base(4,7,0)
+import Data.Typeable (Typeable)
+#endif
+
+-- | A pattern-matching-friendly representation of the signum of a value.
+--
+-- @since 0.7
+data Signum = Negative | Zero | Positive
```

Comment:
In general `signum` can reasonably take more than three values (e. g., for complex numbers there are 5 possibilities). I think it's worth to alert a reader that this is valid only for a sign of an ordered ring. But do we need this type at all? Is not `Ordering` enough?

Reply:
@Bodigrim Actually, could you clarify what those 5 possibilities are? Having looked into it, I see [two generalizations to complex numbers specifically](https://en.wikipedia.org/wiki/Sign_function#Complex_signum), as well as [one even more general one](https://en.wikipedia.org/wiki/Sign_function#Generalized_signum_function) - which of these gives rise to 5 possibilities?

Reply:
I was thinking of Gaussian integers: https://github.com/Bodigrim/arithmoi/blob/master/Math/NumberTheory/Quadratic/GaussianIntegers.hs#L81
Eisenstein integers give even more exotic example, where `signum` takes 7 possible values: https://github.com/Bodigrim/arithmoi/blob/master/Math/NumberTheory/Quadratic/EisensteinIntegers.hs#L90